### ISSUE TITLE
IDE0072: False positive when using switch expression and pattern matching with "or"

### ISSUE THREAD

**Issue by airbreather** [dotnet/roslyn#50982]
**Version Used**: 
.NET SDK 5.0.102

**Steps to Reproduce**:
`dotnet build` the following:
### Program.cs
```csharp
bool EnumIsValid(FriendOrFoe f)
{
    return f switch
    {
        FriendOrFoe.Friend or FriendOrFoe.Foe => true,
        _ => false,
    };
}

enum FriendOrFoe
{
    Friend,
    Foe
}
```

### ConsoleApp0.csproj
```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>net5.0</TargetFramework>
    <EnforceCodeStyleInBuild>true</EnforceCodeStyleInBuild>
  </PropertyGroup>

</Project>
```

### .editorconfig
```editorconfig
[*.cs]
dotnet_diagnostic.IDE0072.severity = warning
```
**Expected Behavior**:
IDE0072 is not raised: all cases are handled

**Actual Behavior**:
IDE0072 is raised.  The suggested fix is to add a separate case for `FriendOrFoe.Friend` and a separate case for `FriendOrFor.Foe`.

**Comment by Youssef1313** [comment]
The following part needs to handle `CaseKind.Pattern`:

https://github.com/dotnet/roslyn/blob/afd10305a37c0ffb2cfb2c2d8446154c68cfa87a/src/Analyzers/Core/Analyzers/PopulateSwitch/PopulateSwitchStatementHelpers.cs#L75-L102

**Comment by airbreather** [comment]
Another report of this was buried in a comment over here: https://github.com/dotnet/roslyn/issues/48876#issuecomment-731208672

**Comment by CyrusNajmabadi** [comment]
Waht i'm not quite understanding is that the above code looks designed to bail out in cases like this and to 'assume that they're complete'.  This seems to not be happening.  We're assuming it's incomplete instead.

**Comment by Youssef1313** [comment]
@CyrusNajmabadi After a second look, I'm confused as well. The code indeed looks designed to assume that it's complete regardless it's complete or not.

I'm currently looking at code only through sourceroslyn.io and GitHub. So can't explain without some debugging (might I've gone through a wrong/irrelevant code path?).

**Comment by Youssef1313** [comment]
Indeed, I went through a wrong code path (which needs to be fixed, but it's for another analyzer).

The part relevant here is the following, which only looks for `IConstantPatternOperation` and ignore patterns:

https://github.com/dotnet/roslyn/blob/afd10305a37c0ffb2cfb2c2d8446154c68cfa87a/src/Analyzers/Core/Analyzers/PopulateSwitch/PopulateSwitchExpressionHelpers.cs#L42-L62
